controllers/bucket_controller.go

```diff
@@ -401,6 +415,123 @@ func (r *BucketReconciler) reconcileWithMinio(ctx context.Context, bucket source
 	return sourcev1.Bucket{}, nil
 }
 
+func (r *BucketReconciler) reconcileWithAzureBlob(ctx context.Context, bucket sourcev1.Bucket, tempDir string) (sourcev1.Bucket, error) {
+	azBlobPipeline, err := r.authAzure(ctx, bucket)
+	if err != nil {
+		err = fmt.Errorf("auth error: %w", err)
+		return sourcev1.BucketNotReady(bucket, sourcev1.AuthenticationFailedReason, err.Error()), err
+	}
+
+	ctxTimeout, cancel := context.WithTimeout(ctx, bucket.Spec.Timeout.Duration)
+	defer cancel()
+
+	ep := strings.TrimRight(bucket.Spec.Endpoint, "/")
+	blobContainerURLString := strings.Join([]string{ep, bucket.Spec.BucketName}, "/")
+	blobContainerURL, err := url.Parse(blobContainerURLString)
+	if err != nil {
+		return sourcev1.Bucket{}, fmt.Errorf("invalid blob container endpoint '%s': %w", blobContainerURLString, err)
+	}
+
+	blobContainer := azblob.NewContainerURL(*blobContainerURL, *azBlobPipeline)
+	if err != nil {
+		return sourcev1.BucketNotReady(bucket, sourcev1.BucketOperationFailedReason, err.Error()), err
+	}
+
+	downloadBlobToLocalFile := func(ctx context.Context, blobPipeline pipeline.Pipeline, url url.URL, blobName, localPath string) error {
+		containerURL := azblob.NewContainerURL(url, blobPipeline)
+		blobURL := containerURL.NewBlobURL(blobName)
+
+		blobProps, err := blobURL.GetProperties(ctx, azblob.BlobAccessConditions{}, azblob.ClientProvidedKeyOptions{})
+		if err != nil {
+			return err
+		}
+
+		content := make([]byte, blobProps.ContentLength())
+		err = azblob.DownloadBlobToBuffer(ctx, blobURL, 0, 0, content, azblob.DownloadFromBlobOptions{})
+		if err != nil {
+			return err
+		}
+
+		dir := filepath.Dir(localPath)
+		err = os.MkdirAll(dir, os.ModePerm)
+		if err != nil {
+			return fmt.Errorf("failed to create local directory '%s': %w", blobURL, err)
+		}
+
+		err = ioutil.WriteFile(localPath, content, 0600)
+		return err
+	}
+
+	// Look for file with ignore rules first
+	path := filepath.Join(tempDir, sourceignore.IgnoreFile)
+
+	if err := downloadBlobToLocalFile(ctxTimeout, *azBlobPipeline, blobContainer.URL(), sourceignore.IgnoreFile, path); err != nil {
+		if resp, ok := err.(azblob.StorageError); ok && resp.ServiceCode() != "BlobNotFound" {
+			return sourcev1.BucketNotReady(bucket, sourcev1.BucketOperationFailedReason, err.Error()), err
+		}
+	}
+
+	ps, err := sourceignore.ReadIgnoreFile(path, nil)
+	if err != nil {
+		return sourcev1.BucketNotReady(bucket, sourcev1.BucketOperationFailedReason, err.Error()), err
+	}
+	// In-spec patterns take precedence
+	if bucket.Spec.Ignore != nil {
+		ps = append(ps, sourceignore.ReadPatterns(strings.NewReader(*bucket.Spec.Ignore), nil)...)
+	}
+
+	matcher := sourceignore.NewMatcher(ps)
+
+	// download blob content
+	marker := azblob.Marker{Val: nil}
+	for true {
```

Comment:
If you want to do an infinite loop it is fine to skip the true.

```go
for {

}
```

Reply:
Done.